ldm/modules/embedding_manager.py

```diff
@@ -169,9 +169,14 @@ def forward(
                     placeholder_embedding.shape[0], max_step_tokens
                 )
 
-                placeholder_rows, placeholder_cols = torch.where(
-                    tokenized_text == placeholder_token.to(device)
-                )
+                if torch.cuda.is_available():
+                    placeholder_rows, placeholder_cols = torch.where(
+                        tokenized_text == placeholder_token.to(device)
+                    )
+                else:
+                    placeholder_rows, placeholder_cols = torch.where(
+                        tokenized_text == placeholder_token
```

Comment:
why? to me it looks like `placeholder_token` is on a different device (CPU), than `tokenized_text` (MPS), so surely `.to(device)` is necessary?

<img width="625" alt="image" src="https://user-images.githubusercontent.com/6141784/192403122-1ef5f959-c967-4691-be1c-fcc20a5b5485.png">

`.to(device)` works fine for me…

Reply:
Are you using `num_vectors_per_token: 6` ? That's where I get an error in that line.

Reply:
Yes, I'm using 6 vectors per token. 1 epoch completed so far:  
https://twitter.com/Birchlabs/status/1574567602722054149

Reply:
@Birch-san the error is after training, when loading the .pt file and then trying to create an image. @tmm1 got an error too. https://github.com/invoke-ai/InvokeAI/issues/517#issuecomment-1258727991 It reads different than mine because he had `lldb` but I'm pretty sure it's the same error (and we had the same solution). See: https://github.com/invoke-ai/InvokeAI/issues/517#issuecomment-1258855834

Reply:
Hope the training goes well!

Reply:
okay _that's_ why you need it...  
![image](https://user-images.githubusercontent.com/6141784/192647166-be48e179-4ef0-4f8b-bb27-f06ac2166104.png)

Reply:
the segfault is caused by:

```python
from torch import cat, ones, tensor, int64
tokenized_text = ones(1, 77, device='mps', dtype=int64)
row = tensor(0, device='mps', dtype=int64)
col = tensor(0, device='mps', dtype=int64)
num_vectors_per_token = 6
placeholder_token = tensor(265, device='mps', dtype=int64)
cat([
  tokenized_text[row][:col],
  placeholder_token.expand(num_vectors_per_token),
  tokenized_text[row][col + 1 :],
], axis=0)
```

or more minimally:

```python
from torch import cat, ones, tensor
cat([
  ones(1, device='mps')[:0],
  tensor([1], device='mps')
], axis=0)
```

seems they have an assertion on nightly to prevent concatting empty tensors though.

I'm gonna filter out empty tensors on my branch:

```python
new_token_row = torch.cat(
    # cat(axis=0) segfaults on MPS if any element has a 0th dim of size 0
    [tensor for tensor in (
        tokenized_text[row][:col],
        placeholder_token.expand(num_vectors_for_token),
        tokenized_text[row][col + 1 :],
    ) if tensor.size(0) > 0],
    axis=0,
)[:n]
new_embed_row = torch.cat(
    # cat(axis=0) segfaults on MPS if any element has a 0th dim of size 0
    [tensor for tensor in (
        embedded_text[row][:col],
        placeholder_embedding[:num_vectors_for_token],
        embedded_text[row][col + 1 :],
    ) if tensor.size(0) > 0],
    axis=0,
)[:n]
```

Reply:
here's the patches I went with, which keeps everything on the same device (GPU):  
https://github.com/Birch-san/stable-diffusion/compare/f5b44a45b7b46c42cf1cf68f61cc5aa3ae10cdae...3d56cc023e36ade51c325fb0d19e1b71976cf236

Reply:
ah, on pytorch stable, it's not even possible to do that select on-GPU; it just does a fallback-to-CPU instead.

```
/Users/birch/git/stable-diffusion/ldm/modules/embedding_manager.py:171: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659484612588/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  placeholder_rows, placeholder_cols = torch.where(
```

is that why your mixed-device approach is working? because it moves them both to CPU anyway?

Reply:
My first intuition, after all the problems we are having on MPS, was to just remove `to(device)`, assuming it was trying to send it to the GPU. Since it worked, I did not explore further, but what you say makes sense. Maybe the operation fails since one operand is not on GPU and falls back to CPU.

Reply:
here's a minimal repro of the problem:  
https://github.com/pytorch/pytorch/issues/86052